File: lab/examples/go/internal/hexagon/domain/order/money.go
package order

import "fmt"

// Money is a value object representing an amount in minor units (e.g. cents)
// together with a 3-letter ISO currency code. It is immutable: operations
// return new values instead of mutating the receiver.
type Money struct {
	amountMinor int64
	currency    string
}

// NewMoney creates a validated Money value.
func NewMoney(amountMinor int64, currency string) (Money, error) {
	if amountMinor < 0 {
		return Money{}, ErrNegativeAmount
	}
	if len(currency) != 3 {
		return Money{}, ErrInvalidCurrency
	}
	return Money{amountMinor: amountMinor, currency: currency}, nil
}

// AmountMinor returns the amount in minor units.
func (m Money) AmountMinor() int64 { return m.amountMinor }

// Currency returns the ISO currency code.
func (m Money) Currency() string { return m.currency }

// Add returns the sum of two Money values of the same currency.
func (m Money) Add(other Money) (Money, error) {
	if m.currency != other.currency {
		return Money{}, ErrCurrencyMismatch
	}
	return Money{amountMinor: m.amountMinor + other.amountMinor, currency: m.currency}, nil
}

// String renders the amount in major units, e.g. "19.99 EUR".
func (m Money) String() string {
	return fmt.Sprintf("%d.%02d %s", m.amountMinor/100, m.amountMinor%100, m.currency)
}

File: lab/examples/kotlin/hexagon/domain/src/main/kotlin/hexslice/hexagon/domain/order/Money.kt
package hexslice.hexagon.domain.order

/**
 * A value object representing an amount in minor units (e.g. cents) together
 * with a 3-letter ISO currency code.
 *
 * Immutable: operations return new values instead of mutating the receiver, and
 * the constructor is private so a Money can only come into existence validated —
 * [of] is the only way in.
 */
@ConsistentCopyVisibility
data class Money private constructor(val amountMinor: Long, val currency: String) {
    /** Returns the sum of two Money values of the same currency. */
    fun add(other: Money): Money {
        if (currency != other.currency) throw OrderError.CurrencyMismatch()
        return Money(amountMinor + other.amountMinor, currency)
    }

    /** Renders the amount in major units, e.g. `19.99 EUR`. */
    override fun toString(): String =
        "${amountMinor / 100}.${(amountMinor % 100).toString().padStart(2, '0')} $currency"

    companion object {
        /** Creates a validated Money value. */
        fun of(amountMinor: Long, currency: String): Money {
            if (amountMinor < 0) throw OrderError.NegativeAmount()
            if (currency.length != 3) throw OrderError.InvalidCurrency()
            return Money(amountMinor, currency)
        }
    }
}
